File: rover.cpp
#include <iostream>
#include <map>
#include <set>
#include <vector>

class Point {
public:
    int x, y;

    Point(int x, int y) : x(x), y(y) {}

    bool operator<(const Point& other) const {
        return std::tie(x, y) < std::tie(other.x, other.y);
    }
};

class Grid {
public:
    int width, height;
    std::set<Point> obstacles;

    Grid(int width, int height) : width(width), height(height) {
        if (width < 0 || height < 0) {
            throw std::invalid_argument("Grid size must be positive.");
        }
    }

    void addObstacle(Point obstacle) {
        obstacles.insert(obstacle);
    }

    bool isObstacle(Point point) {
        return obstacles.count(point) > 0;
    }

    bool isWithinBounds(Point point) {
        return point.x >= 0 && point.x < width && point.y >= 0 && point.y < height;
    }
};

class Rover {
public:
    Point position;
    std::string direction;
    Grid& grid;

    Rover(int x, int y, std::string direction, Grid& grid) : position(x, y), direction(direction), grid(grid) {}

    void move() {
        Point nextPosition = position;
        if (direction == "N") nextPosition.y++;
        else if (direction == "S") nextPosition.y--;
        else if (direction == "E") nextPosition.x++;
        else if (direction == "W") nextPosition.x--;

        if (!grid.isObstacle(nextPosition)) {
            if (grid.isWithinBounds(nextPosition)) {
                position = nextPosition;
            } else {
                std::cout << "Error: Rover cannot move outside the grid.\n";
            }
        } else {
            throw std::runtime_error("Rover cannot move to an obstacle.");
        }
    }

    void turnLeft() {
        if (direction == "N") direction = "W";
        else if (direction == "S") direction = "E";
        else if (direction == "E") direction = "N";
        else if (direction == "W") direction = "S";
    }

    void turnRight() {
        if (direction == "N") direction = "E";
        else if (direction == "S") direction = "W";
        else if (direction == "E") direction = "S";
        else if (direction == "W") direction = "N";
    }

    void executeCommand(char command) {
        if (command == 'M') move();
        else if (command == 'L') turnLeft();
        else if (command == 'R') turnRight();
    }

    void printStatus() {
        std::cout << "Rover Status:\n";
        std::cout << "Position: (" << position.x << ", " << position.y << ")\n";
        std::cout << "Direction: " << direction << "\n";
    }
};

int main() {
    int x, y;
    std::string direction;
    std::cout << "Enter the starting position of the rover (x, y, direction):\n";
    std::cin >> x >> y >> direction;

    int width, height;
    std::cout << "Enter the size of the grid (width, height):\n";
    std::cin >> width >> height;

    Grid grid(width, height);

    int numObstacles;
    std::cout << "Enter the number of obstacles:\n";
    std::cin >> numObstacles;

    for (int i = 0; i < numObstacles; i++) {
        std::cout << "Enter the coordinates of obstacle " << (i + 1) << " (x, y):\n";
        std::cin >> x >> y;
        grid.addObstacle(Point(x, y));
    }

    Rover rover(x, y, direction, grid);

    std::cout << "Enter the commands for the rover:\n";
    std::string commands;
    std::cin >> commands;

    for (char command : commands) {
        rover.executeCommand(command);
    }

    rover.printStatus();

    return 0;
}

File: rover.java
import java.util.*;

interface Command {
    void execute();
}

class Rover {
    int x, y;
    String direction;
    Map<String, Command> commands = new HashMap<>();

    Rover(int x, int y, String direction) {
        this.x = x;
        this.y = y;
        this.direction = direction;
        commands.put("M", this::move);
        commands.put("L", this::turnLeft);
        commands.put("R", this::turnRight);
    }

    void move() {
        if (direction.equals("N"))
            y++;
        else if (direction.equals("S"))
            y--;
        else if (direction.equals("E"))
            x++;
        else if (direction.equals("W"))
            x--;
    }

    void turnLeft() {
        if (direction.equals("N"))
            direction = "W";
        else if (direction.equals("S"))
            direction = "E";
        else if (direction.equals("E"))
            direction = "N";
        else if (direction.equals("W"))
            direction = "S";
    }

    void turnRight() {
        if (direction.equals("N"))
            direction = "E";
        else if (direction.equals("S"))
            direction = "W";
        else if (direction.equals("E"))
            direction = "S";
        else if (direction.equals("W"))
            direction = "N";
    }

    void executeCommand(String command) {
        commands.get(command).execute();
    }
}

class Grid {
    int width, height;
    Set<Point> obstacles = new HashSet<>();

    Grid(int width, int height) {
        this.width = width;
        this.height = height;
    }

    void addObstacle(Point obstacle) {
        obstacles.add(obstacle);
    }

    boolean isObstacle(Point point) {
        return obstacles.contains(point);
    }
}

class Point {
    int x, y;

    Point(int x, int y) {
        this.x = x;
        this.y = y;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o)
            return true;
        if (o == null || getClass() != o.getClass())
            return false;
        Point point = (Point) o;
        return x == point.x && y == point.y;
    }

    @Override
    public int hashCode() {
        return Objects.hash(x, y);
    }
}

public class rover {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        System.out.println("Enter the starting position of the rover (x, y, direction):");
        int x = scanner.nextInt();
        int y = scanner.nextInt();
        String direction = scanner.next();

        Rover rover = new Rover(x, y, direction);

        System.out.println("Enter the size of the grid (width, height):");
        int width = scanner.nextInt();
        int height = scanner.nextInt();

        Grid grid = new Grid(width, height);

        System.out.println("Enter the number of obstacles:");
        int numObstacles = scanner.nextInt();

        for (int i = 0; i < numObstacles; i++) {
            System.out.println("Enter the coordinates of obstacle " + (i + 1) + " (x, y):");
            int obstacleX = scanner.nextInt();
            int obstacleY = scanner.nextInt();
            grid.addObstacle(new Point(obstacleX, obstacleY));
        }

        System.out.println("Enter the commands for the rover:");
        String[] commands = scanner.next().split("");

        for (String command : commands) {
            Point nextPoint = new Point(rover.x, rover.y);
            if (command.equals("M")) {
                if (rover.direction.equals("N"))
                    nextPoint.y++;
                else if (rover.direction.equals("S"))
                    nextPoint.y--;
                else if (rover.direction.equals("E"))
                    nextPoint.x++;
                else if (rover.direction.equals("W"))
                    nextPoint.x--;
            }

            if (!grid.isObstacle(nextPoint)) {
                rover.executeCommand(command);
            }
        }

        System.out.println("Final Position: (" + rover.x + ", " + rover.y + ", " + rover.direction + ")");
    }
}